### PR TITLE
ARROW-10942: [C++] Fix S3FileSystem::Impl::IsEmptyDirectory on Amazon

### PR DESCRIPTION
**Pull request by pitrou**
Amazon and Minio S3 implementations unfortunately have slightly different conventions for recognizing and handling empty "directories".

### CODE REVIEW

**Review by juanjgalvez — COMMENTED** [review]
Wow, thanks for the quick resolution! This looks good to me. Would it make sense to do something similar in `IsNonEmptyDirectory`? I have not observed issue there myself, but we have a user organization that implements their own S3-compatible object store and they were having related issues with GetFileInfo when applied to a directory. In their case the issue was that arrow could not determine if it was a file or a directory (result was unknown). I have no simple way to debug this right now on their end.

**Comment by juanjgalvez** [comment]
Btw, tested it and it works.

**Review by juanjgalvez — APPROVED** [review]

File: cpp/src/arrow/filesystem/s3fs.cc [hunk 1]
```diff
@@ -1241,6 +1245,13 @@ class S3FileSystem::Impl {
     return std::string(FromAwsString(builder_.config().region));
   }
 
+  template <typename Error>
+  void SaveBackend(const Aws::Client::AWSError<Error>& error) {
+    if (!backend_ || *backend_ == S3Backend::Other) {
```
Comment (kou) [inline cpp/src/arrow/filesystem/s3fs.cc:1250]: Do we need this condition?

**Review by kou — APPROVED** [review]
+1

File: cpp/src/arrow/filesystem/s3fs.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/filesystem/s3fs.cc:1250]: Not necessarily, but if the backend was already detected, we probably don't want to undo the detection if the HTTP header sometimes misses or is incorrect.

**Comment by pitrou** [comment]
> Would it make sense to do something similar in IsNonEmptyDirectory?

I don't think so. I doesn't use the same S3 API call so so will be subject to different conditions. But if you find a bug with it, feel free to report it.

**Pull request closed without merging**